components/suggestion/command_provider/app_command_parser/app_command_parser.ts

```diff
@@ -0,0 +1,1256 @@
+                            defaultMessage: 'Matching tick quote expected before end of input.',
+                        }));
+                    }
+                    return this;
+                }
+                case '`': {
+                    if (this.incompleteStart === this.i - 1) {
+                        return this.asError(this.intl.formatMessage({
+                            id: 'apps.error.parser.empty_value',
+                            defaultMessage: 'empty values are not allowed',
+                        }));
+                    }
+                    this.i++;
+                    this.state = ParseState.EndTickedValue;
+                    break;
+                }
+                default: {
+                    this.incomplete += c;
+                    this.i++;
+                    break;
+                }
+                }
+                break;
+            }
+
+            case ParseState.EndTickedValue:
+            case ParseState.EndQuotedValue:
+            case ParseState.EndValue: {
+                if (!this.field) {
+                    return this.asError(this.intl.formatMessage({
+                        id: 'apps.error.parser.missing_field_value',
+                        defaultMessage: 'Field value Expected.',
+                    }));
+                }
+
+                // special handling for optional BOOL values ('--boolflag true'
+                // vs '--boolflag next-positional' vs '--boolflag
+                // --next-flag...')
+                if (this.field.type === AppFieldTypes.BOOL &&
+                    ((autocompleteMode && !'true'.startsWith(this.incomplete) && !'false'.startsWith(this.incomplete)) ||
+                    (!autocompleteMode && this.incomplete !== 'true' && this.incomplete !== 'false'))) {
+                    // reset back where the value started, and treat as a new parameter
+                    this.i = this.incompleteStart;
+                    this.values![this.field.name] = 'true';
+                    this.state = ParseState.StartParameter;
+                } else {
+                    if (autocompleteMode && c === '') {
+                        return this;
+                    }
+                    this.values![this.field.name] = this.incomplete;
+                    this.incomplete = '';
+                    this.incompleteStart = this.i;
+                    if (c === '') {
+                        return this;
+                    }
+                    this.state = ParseState.ParameterSeparator;
+                }
+                break;
+            }
+            }
+        }
+    }
+}
+
+export class AppCommandParser {
+    private store: Store;
+    private channelID: string;
+    private rootPostID?: string;
+    private intl: Intl;
+
+    forms: {[location: string]: AppForm} = {};
+
+    constructor(store: Store|null, intl: Intl, channelID: string, rootPostID = '') {
+        this.store = store || getStore() as Store;
```

Comment:
(1/5) When would the provided store be null? Perhaps we could always pass in the store to remove a reference to the global instance of the store?

Reply:
This `app_command_parser.ts` file is used in the mobile app as well. The store is not being passed in the [mobile app's case](https://github.com/mattermost/mattermost-mobile/blob/c8f8d2c35c78095044c21049667e5b188942743b/app/components/autocomplete/slash_suggestion/slash_suggestion.tsx#L69) because it is being instantiated in a React component. This is why the app command parser needs to handle this case. What are your thoughts on this?

Sorry for not explaining the shared code between the webapp and the mobile app earlier. The `app_command_parser_dependencies.ts` file is meant to be the one difference between the two platforms' implementations. Especially since the mobile app's imports of `mattermost-redux` changed to `@mm-redux`, this needed to be done this way.

Reply:
Ah, that makes sense. If we ever decide to isolate this from the global store then, it'd be easy enough to just have `getStore` return undefined anyway since we're always passing in a value in the web app. For now though, I'm fine leaving it this way.

As for the dependencies, I didn't realize that was for mobile support. I figured it was actually to make mocking out the dependencies easier for testing since that also seems like a good case for it. Having to partially mock a module always ends up being a pain in the neck, so doing it this way does make it easier to mock only what you use.